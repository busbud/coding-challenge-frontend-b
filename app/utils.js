// UTC time (dateTtime) -> hh:mm
export const utcTimeToHhMm = utcTime => utcTime.split('T')[1].substr(0, 5)

// Number XXXX -> XX,XX (currency)
export const numToCurrency = num => (num/100).toFixed(2)
