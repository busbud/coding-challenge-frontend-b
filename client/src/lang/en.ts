export default {
  title: "Find bus tickets for your trip",
  chooseCurrency: "Choose Currency",
  chooseLanguage: "Choose Language",
  origin: "Origin",
  destination: "Destination",
  date: "Date",
  passenger: "Passenger",
  submit: "Submit",
  passengerUnit: "passenger",
  resultList: "Result List",
};
