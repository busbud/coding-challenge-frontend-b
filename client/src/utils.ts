export const toCamel = (s: string) => {
  return s.replace(/([-_][a-z])/gi, ($1) => {
    return $1.toUpperCase().replace("-", "").replace("_", "");
  });
};

export const capitalize = (str: string) => {
  const lower = str.toLowerCase();
  return str.charAt(0).toUpperCase() + lower.slice(1);
};
