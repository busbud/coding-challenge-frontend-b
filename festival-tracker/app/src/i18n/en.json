{
  "Loading...": "Loading...",

  "No departures found": "No departures found",
  "Find the best bus ticket to your destination !": "Find the best bus tickets to your destination !",
  "Date": "Date",
  "Origin": "Origin",
  "Select an origin": "Select an origin",
  "Destination": "Destination",
  "Select a destination": "Select a destination",
  "Adult": "Adult",
  "Child": "Child",
  "Senior (65+)": "Senior (65+)",
  "Country": "Country",
  "Select a country": "Select a country", 
  "Currency": "Currency",
  "Select a currency": "Select a currency",
  "Search": "Search",
  "Results": "Results",
  "Previous": "Previous",
  "Next": "Next",

  "Departure": "Departure",
  "Arrival": "Arrival",
  "Trip": "Trip",
  "Price": "Price"
}