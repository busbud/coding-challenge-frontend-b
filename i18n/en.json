{
  "LABEL_FROM": "From",
  "LABEL_TO": "To",
  "LABEL_DEPART": "Depart",
  "LABEL_TRAVELLERS": "Travellers",
  "INPUT_CITY": "City",
  "LABEL_PASSENGERS": "Passengers",
  "INPUT_PASSENGER": "passenger",
  "INPUT_ADULTS": "Adults",
  "INPUT_CHILDREN": "Children",
  "INPUT_SENIORS": "Seniors",
  "BUTTON_SEARCH": "Search for buses",
  "PROMOTION_TITLE": "On your way",
  "PROMOTION_TRIP": "Travelling from New York",
  "PROMOTION_TRIP_DATE": "Sep 28",
  "PROMOTION_DATE": "29 Sept",
  "PROMOTION_URL": "/results?lang=en&origin=New+York&destination=Montreal&departure=28-09-2018&adults=1&children=0&seniors=0",
  "PROMOTION_BUTTON": "Join to the party!",
  "ERROR_RESULT": "No results found",
  "ERROR_SEARCH": "Search failed"
}
