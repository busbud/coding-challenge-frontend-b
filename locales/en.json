{
  "welcome" : "Get your bus ticket to Osheaga with Busbud",
  "cta" : "Book yours Now",
  "home-link" : "Home",
  "language-link" : "Language",
  "book-now-link" : "Book Now"
}