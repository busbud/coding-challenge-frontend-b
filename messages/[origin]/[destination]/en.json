{
  "Search": {
    "description": "Departures available",
    "emptyState.title": "No available trips on your selected date",
    "emptyState.description": "None of our partners are operating trips for your chosen date. We apologize for this inconvenience."
  }
}