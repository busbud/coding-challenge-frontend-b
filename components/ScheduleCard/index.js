import React, { useContext, useState } from "react";
import styled from "styled-components";

import { IntlContext } from "../../pages/_app";
import { getTransaltion } from "../../utils/translation";
import { formatedDate, formatedTime, findDifference } from "../../utils/date";
import Arrow from "../Arrow";
import { Row, Column } from "../../components/common-styled/Containers";
import { BoldText, NormalText } from "../../components/common-styled/Texts";
import { media } from "../common-styled/device";

const Wrapper = styled(Column)`
  border: 1px solid #ccc;
  padding: 16px;
  border-radius: 5px;
  flex-direction: column;
  margin: 10px 0;
`;
const Box = styled(Column)`
  align-items: center;
  justify-content: center;
  padding: 20px 40px;
  border-radius: 10px;
  border: 1px solid #ccc;
  ${media.mobile`
    ${props => props.alignEnd && "align-self: flex-end;"}
  `}
`;
const TopRow = styled(Row)`
  ${media.mobile`flex-direction: column;`}
`;
const BottomRow = styled(Row)`
  ${media.mobile`
    flex-direction: column;
    align-items: flex-start;
  `}
  align-items: center;
`;
const City = styled.div`
  color: gray;
  margin-top: 5px;
`;
const OperatorLogo = styled.img``;
const OperatorName = styled.div`
  margin-left: 10px;
`;
const Price = styled(BoldText)`
  margin: 0 10px 0 auto;
  ${media.mobile`margin: 0;`}
`;
const Book = styled.button`
  background: #065af3;
  border-radius: 96px;
  color: #fff;
  font-size: 14px;
  padding: 10px 20px;
  text-align: center;
  outline: none;
  cursor: pointer;
  ${media.mobile`margin: 0 0 0 auto;`}
`;
const Date = styled.div`
  margin-left: 10px;
  border-left: 1px solid #9e9a9a;
  padding-left: 10px;
  ${media.mobile`display: none;`}
`;
const ArrowSpaced = styled(Arrow)`
  margin: 15px 30px;
  ${media.mobile`align-self: center;`}
`;
const DetailsArrow = styled(Arrow)`
  margin: -1px 0 0 10px;
`;
const Details = styled.a`
  margin: auto 0 0 auto;
  display: flex;
  align-items: center;
  color: #1da1f2;
  cursor: pointer;
  ${media.mobile`margin: 16px 0 0 auto;`}
`;
const Expandable = styled(Row)`
  ${props => (props.isOpen ? "max-height:1000px;" : "max-height:0;")}
  transition: max-height 0.3s ease-out;
  margin-top: 10px;
  overflow: hidden;
`;
const ExtrasWrap = styled(Row)`
  padding: 10px;
  border: 1px solid #ccc;
  flex-grow: 1;
  border-radius: 5px;
`;
const Amenities = styled(Row)`
  margin: 15px 0 0 -10px;
  flex-wrap: wrap;
`;
const Amenity = styled(Book)`
  margin: 0 10px;
  ${media.desktop`
    margin: 0 0 10px 10px;
  `}
`;
const MobileSection = styled.div`
  display: flex;
  flex-grow: 1;
  align-items: center;
  margin-bottom: 16px;
`;

export default function ScheduleCard(props) {
  const { language } = useContext(IntlContext);
  const { schedule, operator = {}, origin = {}, destination = {} } = props;
  const [isOpen, setOpen] = useState(false);
  const { mins, hrs } = findDifference(
    schedule.departure_time,
    schedule.arrival_time
  );
  return (
    <Wrapper key={schedule.id}>
      <TopRow>
        <MobileSection>
          <OperatorLogo
            alt={operator.display_name}
            src={operator.logo_url}
          ></OperatorLogo>
          <OperatorName data-testid="scard-operator">
            {operator.display_name}
          </OperatorName>
          <Date data-testid="scard-departure-date">
            {formatedDate(schedule.departure_time)}
          </Date>
        </MobileSection>
        <MobileSection>
          <Price data-testid="scard-price">
            {schedule.prices.currency +
              " " +
              (schedule.prices.total / 100).toFixed(2)}
          </Price>
          <Book>BOOK NOW</Book>
        </MobileSection>
      </TopRow>
      <BottomRow>
        <Box>
          <BoldText data-testid="scard-departure">
            {formatedTime(schedule.departure_time)}
          </BoldText>
          <City>{origin.name}</City>
        </Box>
        <ArrowSpaced
          dotted
          direction={"right"}
          text={(hrs && hrs + "hrs ") + (mins && mins + "mins")}
          bottomText={hrs > 24 && "+1 day"}
        />
        <Box alignEnd>
          <BoldText>{formatedTime(schedule.arrival_time)}</BoldText>
          <City>{destination.name}</City>
        </Box>
        <Details
          data-testid="scard-show-details"
          onClick={() => setOpen(!isOpen)}
        >
          {getTransaltion("sc.bus_details", language)}
          <DetailsArrow direction={"down"} />
        </Details>
      </BottomRow>
      {/* only render if isOpen true */}
      {isOpen && (
        <Expandable isOpen={isOpen}>
          <ExtrasWrap>
            <Column>
              <BoldText data-testid="scard-amenities">
                {getTransaltion("amenities", language)}
              </BoldText>
              <Amenities>
                {Object.keys(schedule.amenities).map((amenity, index) => {
                  return (
                    schedule.amenities[amenity] &&
                    getTransaltion("amenities." + amenity, language) && (
                      <Amenity key={index + "amenity"}>
                        {getTransaltion("amenities." + amenity, language)}
                      </Amenity>
                    )
                  );
                })}
              </Amenities>
            </Column>
          </ExtrasWrap>
        </Expandable>
      )}
    </Wrapper>
  );
}
