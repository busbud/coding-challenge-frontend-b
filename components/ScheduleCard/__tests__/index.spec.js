import React from "react";
import { fireEvent, waitForElement } from "@testing-library/react";

import ScheduleCard from "../index";
import { setupComponent } from "../../../utils/jest";
import mockSchedules from "../../../__mocks__/schedules";
import { formatedDate, formatedTime } from "../../../utils/date";
import en from "../../../translations/en";

describe("ScheduleCard component", () => {
  it("should render without throwing an error", function() {
    const { asFragment } = setupComponent(
      <ScheduleCard
        schedule={mockSchedules.departures[0]}
        operator={mockSchedules.operators[0]}
        origin={mockSchedules.locations[0]}
        destination={mockSchedules.locations[1]}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render data as receveid", function() {
    const schedule = mockSchedules.departures[0];
    const operator = mockSchedules.operators[0];
    const { getByTestId } = setupComponent(
      <ScheduleCard
        schedule={schedule}
        operator={operator}
        origin={mockSchedules.locations[0]}
        destination={mockSchedules.locations[1]}
      />
    );
    expect(getByTestId("scard-operator")).toHaveTextContent(
      operator.display_name
    );
    expect(getByTestId("scard-departure-date")).toHaveTextContent(
      formatedDate(schedule.departure_time)
    );
    expect(getByTestId("scard-price")).toHaveTextContent(
      schedule.prices.currency + " " + (schedule.prices.total / 100).toFixed(2)
    );
    expect(getByTestId("scard-departure")).toHaveTextContent(
      formatedTime(schedule.departure_time)
    );
  });

  it("should show bus details only after click", async function() {
    const { getByTestId, debug } = setupComponent(
      <ScheduleCard
        schedule={mockSchedules.departures[0]}
        operator={mockSchedules.operators[0]}
        origin={mockSchedules.locations[0]}
        destination={mockSchedules.locations[1]}
      />
    );

    fireEvent.click(getByTestId("scard-show-details"));
    const amenitiesNode = await waitForElement(() =>
      getByTestId("scard-amenities")
    );
    expect(amenitiesNode).toHaveTextContent(en.amenities);
  });
});
