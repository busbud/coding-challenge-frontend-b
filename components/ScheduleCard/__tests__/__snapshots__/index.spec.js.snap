// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScheduleCard component should render without throwing an error 1`] = `
<DocumentFragment>
  <div
    class="sc-dnqmqq sc-VigVT kplqdz"
  >
    <div
      class="sc-htoDjs sc-fjdhpX jUwZXC"
    >
      <div
        class="sc-iAyFgw dMZbes"
      >
        <img
          class="sc-kAzzGY jHprQs"
          src="https://busbud.imgix.net/operator-logos/logo_trailways.png?h={height}&w={width}&auto=format&fit=fill&bg=0FFF"
        />
        <div
          class="sc-chPdSV fVMEOF"
          data-testid="scard-operator"
        >
          Adirondack Trailways
        </div>
        <div
          class="sc-kpOJdX hMXPOi"
          data-testid="scard-departure-date"
        >
          August 2, 2020
        </div>
      </div>
      <div
        class="sc-iAyFgw dMZbes"
      >
        <div
          class="sc-gZMcBi sc-kgoBCf eJuSOi"
          data-testid="scard-price"
        >
          USD 78.50
        </div>
        <button
          class="sc-kGXeez bJStZD"
        >
          BOOK NOW
        </button>
      </div>
    </div>
    <div
      class="sc-htoDjs sc-jzJRlG geYYoV"
    >
      <div
        class="sc-dnqmqq sc-jTzLTM eCLnUn"
      >
        <div
          class="sc-gZMcBi hmYReZ"
          data-testid="scard-departure"
        >
          11:00 AM
        </div>
        <div
          class="sc-cSHVUG ZOTVl"
        >
          New York City
        </div>
      </div>
      <div
        class="sc-bdVaJa bPGLpq sc-dxgOiQ kkhrFI"
        direction="right"
      >
        <div
          class="sc-bwzfXH kKwxuZ"
        >
          8hrs 55mins
        </div>
        <div
          class="sc-htpNat kxRvrj"
        >
          <div
            class="sc-bxivhb vBGNy"
          />
          <div
            class="sc-ifAKCX bsLoLO"
            direction="right"
          />
        </div>
      </div>
      <div
        class="sc-dnqmqq sc-jTzLTM gtcKXT"
      >
        <div
          class="sc-gZMcBi hmYReZ"
        >
          7:55 PM
        </div>
        <div
          class="sc-cSHVUG ZOTVl"
        >
          31st St & 8th Ave
        </div>
      </div>
      <a
        class="sc-jKJlTe flLZUW"
        data-testid="scard-show-details"
      >
        Bus details
        <div
          class="sc-bdVaJa bPGLpq sc-ckVGcZ eugcqh"
          direction="down"
        >
          <div
            class="sc-bwzfXH kKwxuZ"
          />
          <div
            class="sc-htpNat kxRvrj"
          >
            <div
              class="sc-ifAKCX xeDUD"
              direction="down"
            />
          </div>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
